### PR TITLE
Testing/test runner refactor

### PR DESCRIPTION
**Pull request by embray**
Following discussion in #2317, here's some changes that will make it I think a little easier to add changes to the `astropy.tests` test runner.  This also moves the `AstropyTest` distutils command class back into astropy.tests so that it isn't awkwardly separated in astropy-helpers.  There will be an associated change in astropy-helpers, so this PR should include a submodule update before it's merged.

This PR depends on #4017.

### CODE REVIEW

**Comment by Cadair** [comment]
so sunpy has a subclass of the `AstropyTest` class, in which I have defined our own arguments, mainly to fiddle the online offline options. I am assuming that this, and the respective changes in `astropy_helpers` would mean that our test command would have to import astropy first??

https://github.com/sunpy/sunpy/blob/master/sunpy/tests/setup_command.py

**Comment by embray** [comment]
@Cadair That's right.  You won't have to make that change immediately though.  The old command class will be kept for backwards compat for now.

Maybe instead though you could describe what you mean by "fiddle the online offline options".  Could that just be included in the test runner?

**Comment by Cadair** [comment]
So, when I moved sunpy over to ah the testing infrastructure was a bit of a botch, I wanted to keep our previous online mark behaviour from pytest i.e. `-k online / -k-online` so I rigged up a `--online` and `--online-only` flag in the TestHelper https://github.com/sunpy/sunpy/blob/master/sunpy/tests/setup_command.py#L30.

I was discussing with @bsipocz that our runner probably needs re-writing because we seem to be having some issues with the doctest build stalling on travis, and as I said it's a bit of a hack.

How would I change these options from the test runner?

**Comment by bsipocz** [comment]
@Cadair - or there may be no need to rewrite stuff if it will be easy to just add extra options (like the planned ``--slow``). You could just patch ``--online`` on top of the ah/astropy test suite and the rest would work as is. 

Thus I guess all the astropy doctest directives would work out of the box and solve sunpy/sunpy#1456 and the halting builds, etc.

**Comment by embray** [comment]
This should make it easier to add new options, yes.

On the astropy-helpers end of things, it will work so that even if you don't have astropy available there is still a dummy "test" command, so users can see that it's available.  It just won't work if astropy isn't installed (and it says as much if you try to use it).

Also, @mdboom and I were talking possibly at some point pulling the astropy testing helpers out to a separate package (not part of astropy-helpers, but something else).  I don't have an ETA on that though.

File: astropy/__init__.py [hunk 1]
```diff
@@ -113,96 +112,9 @@ class Conf(_config.ConfigNamespace):
...
-    test_runner = _get_test_runner()
-    return test_runner.run_tests(
-        package=package, test_path=test_path, args=args,
-        plugins=plugins, verbose=verbose, pastebin=pastebin,
-        remote_data=remote_data, pep8=pep8, pdb=pdb,
-        open_files=open_files, parallel=parallel, docs_path=docs_path,
-        skip_docs=skip_docs, repeat=repeat)
+# Create the test() function
+from .tests.runner import TestRunner
+test = TestRunner.make_test_runner_in(__path__[0])
```
Comment (embray) [inline astropy/__init__.py]: For the affiliated package template we may need something *slightly* ugly here along the lines of:

```python
try:
    from .tests.runner import TestRunner
    test = TestRunner.make_test_runner_in(__path__[0])
except ImportError:
    def test(*args, **kwargs):
        """Run the package test suit.  Requires the astropy package to be installed."""
        raise RuntimeError("The astropy package must be installed in order to run the tests.")
```

**Comment by Cadair** [comment]
So how does this make it easier to add new options to the TestHelper for flags to setup.py?

**Comment by embray** [comment]
Did you mean "for" or "or"?  

Adding options to setup.py is still treated independently from the TestRunner class, for a number of reasons.  But by being in the same package it means all options added to the `test` command are supported and implemented in the TestRunner.  Since the `test` command is just a thin wrapper around TestRunner they're intrinsically coupled and belong together.

**Comment by embray** [comment]
In other words, it's "easier" in the sense that you only have to make changes in one or two places (all in `astropy.tests`) and don't have to orchestrate a corresponding change in astropy-helpers, and then have to make sure that the correct version of astropy-helpers is always used to support whatever options you're adding.

If you want to subclass you just have to subclass `astropy.tests.runner.TestRunner` and `astropy.tests.command.AstropyTest`, and you can rest assured that those classes are in sync with each other as far as what features they support.  In retrospect it was a mistake to separate those classes (and only those classes) into separate packages.

**Comment by embray** [comment]
If there are no objections, this should be included in v1.1.0.  To be clear: this doesn't have any effect in terms of backwards-compatibility.

**Comment by embray** [comment]
[rebased on master]

**Comment by embray** [comment]
Hmm, not clear *at all* why this blew up so badly.  Need to rebase again, then we'll see...

**Comment by astrofrog** [comment]
@embray - can you rebase? Do you want to get this into 1.1?

**Comment by embray** [comment]
Rebased and fixed indentation problem.

**Comment by embray** [comment]
Wow, this is finally ready to merge \o/

Anyone want to review this before I do?

File: astropy/tests/__init__.py [hunk 2]
```diff
@@ -4,8 +4,3 @@
 for writing tests, and general tests that are not associated with a
 particular package.
 """
-from __future__ import (absolute_import, division, print_function,
-                        unicode_literals)
-
-from . import helper
-from .disable_internet import turn_off_internet, turn_on_internet
```
Comment (astrofrog) [inline astropy/tests/__init__.py]: Should we consider this as being part of the public API and therefore keep the last import?

**Comment by astrofrog** [comment]
Apart from my small comment above, this is good to go.

File: astropy/tests/__init__.py (hunk 2 above)
Comment (embray) [inline astropy/tests/__init__.py]: No, it's never been documented as such.  I don't think this import being here was ever particularly deliberate, and nobody seems to rely on it.

**Pull request merged**